### ISSUE TITLE
[QST] Can we improve performance of Parquet file scans with large string columns?

### ISSUE THREAD

**Issue by andygrove** [rapidsai/cudf#7008]
**What is your question?**

We are seeing poor performance in Spark on GPU for Parquet file scans where the files contain columns of large strings. Performance is considerably slower than Spark running on CPU in this case.

In order to reproduce this issue in cuDF, I wrote a data generator and benchmark in Python and the source code is included below.

Two files are generated - `small-strings.parquet` and `large-strings.parquet`.

Both files have two columns where:

- c0 is a 32 byte string
- c1 is also a string but with different sizes between the two files (32 bytes vs 4480 bytes)

There files are generated with different row counts to ensure that the overall volume of data is the same (although the final file size does vary due to compression differences).

I am seeing a 3x difference in file scan performance between these two files, with `large-strings.parquet` being slower.

cuDF is ~35x faster than Pandas for the small-string case but "only" 1.4x faster for the large-string case

## output from running datagen.py

```
Generating 224000 rows x 4480 bytes; total size = 1003520000 (large-strings.parquet)
Generating 15680000 rows x 64 bytes; total size = 1003520000 (small-strings.parquet)
```

## file sizes

```
998472247 large-strings.parquet
619507382 small-strings.parquet
```

## benchmark results

```
Read /tmp/small-strings.parquet in 2.294814109802246
Read /tmp/small-strings.parquet in 0.26930928230285645
Read /tmp/small-strings.parquet in 0.2762606143951416
Read /tmp/small-strings.parquet in 0.2749598026275635
Read /tmp/small-strings.parquet in 0.285076379776001
Read /tmp/small-strings.parquet in 0.27698278427124023
```

```
Read /tmp/large-strings.parquet in 2.949741840362549
Read /tmp/large-strings.parquet in 0.9218177795410156
Read /tmp/large-strings.parquet in 0.9389686584472656
Read /tmp/large-strings.parquet in 0.9387798309326172
Read /tmp/large-strings.parquet in 0.9365167617797852
Read /tmp/large-strings.parquet in 0.9358391761779785
```

## datagen.py

```python
import cudf as pd
import random
import string

# filename = '/tmp/large-strings.parquet'
# rows = 224000
# str_len = 4448

filename = '/tmp/small-strings.parquet'
rows = 224000 * 70
str_len = 32

row_size = 32 + str_len
total_bytes = rows * row_size

print("Generating {} rows x {} bytes; total size = {}".format(rows, row_size, total_bytes))
c0 = []
c1 = []
for i in range(0, rows):
    str = "{:32d}".format(i)
    random_string = ''.join(random.choices(string.ascii_letters, k=str_len))
    c0.append(str)
    c1.append(random_string)

c0_series = pd.Series(c0)
c1_series = pd.Series(c1)
df = pd.DataFrame({ 'c0': c0_series, 'c1': c1_series })
df.to_parquet(filename)
```

## bench.py

```python
import cudf as pd
import time

# filename = "/tmp/small-strings.parquet"
filename = "/tmp/large-strings.parquet"

for i in range(0, 6):
    start = time.time()
    df = pd.read_parquet(filename)
    end = time.time()
    print("Read {} in {}".format(filename, end - start))
```

**Comment by OlivierNV** [comment]
[yes]

**Comment by devavret** [comment]
Possibly be fixed by #7576 . Need to confirm if that conversion is indeed the bottleneck.

**Comment by devavret** [comment]
#7576 has been merged. @andygrove Can you check if the issue persists?

**Comment by andygrove** [comment]
Thanks for the ping @devavret. I asked @viadea if he could verify the fix.

**Comment by viadea** [comment]
Used Andy's tool on 0.19-nightly cudf:
`$  python bench.py
Read /tmp/small-strings.parquet in 1.44537353515625
Read /tmp/small-strings.parquet in 0.2352135181427002
Read /tmp/small-strings.parquet in 0.2478954792022705
Read /tmp/small-strings.parquet in 0.24788236618041992
Read /tmp/small-strings.parquet in 0.24469709396362305
Read /tmp/small-strings.parquet in 0.24921441078186035
Read /tmp/large-strings.parquet in 0.23854398727416992
Read /tmp/large-strings.parquet in 0.235579252243042
Read /tmp/large-strings.parquet in 0.23622751235961914
Read /tmp/large-strings.parquet in 0.2348644733428955
Read /tmp/large-strings.parquet in 0.23695826530456543
Read /tmp/large-strings.parquet in 0.2357475757598877`

They are the same now.

**Comment by devavret** [comment]
Closed by #7576
